The OpenAPI input base directory does not exist

### Tell us the extension you're using

Server

### I tried this:

**I added the quarkus-openapi-generator-server plugin to my quarkus project.**

`dependencies {
    implementation 'io.quarkiverse.openapi.generator:quarkus-openapi-generator-server:2.9.1-lts'
    implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
    implementation 'io.quarkus:quarkus-vertx'
    implementation 'io.quarkus:quarkus-rest-jackson'
    implementation 'io.quarkus:quarkus-arc'
    implementation("io.smallrye.reactive:smallrye-mutiny-vertx-web-client")
    implementation("io.quarkus:quarkus-reactive-pg-client")
    testImplementation 'io.quarkus:quarkus-junit5'
}`

**I placed my openapi spec in a yaml file:**
src/main/resources/openapi/security.yaml

**I configured the yaml file:**
quarkus.openapi.generator.spec=security.yaml

**ran ./gradlew build:**

`Caused by: io.quarkus.bootstrap.prebuild.CodeGenException: The OpenAPI input base directory does not exist. Please create the directory at /Users/sg/dev/quarkus-getting-started-vertx/build/classes/java/quarkus-generated-sources/resources
        at io.quarkiverse.openapi.server.generator.deployment.codegen.ApicurioOpenApiServerCodegen.validateOpenApiDir(ApicurioOpenApiServerCodegen.java:102)
        at io.quarkiverse.openapi.server.generator.deployment.codegen.ApicurioOpenApiServerCodegen.trigger(ApicurioOpenApiServerCodegen.java:61)
        at io.quarkus.deployment.CodeGenerator.lambda$trigger$4(CodeGenerator.java:206)
        at io.quarkus.deployment.CodeGenerator.callWithClassloader(CodeGenerator.java:181)
        at io.quarkus.deployment.CodeGenerator.trigger(CodeGenerator.java:203)
        at io.quarkus.deployment.CodeGenerator.initAndRun(CodeGenerator.java:80)
        at java.base@21/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        ... 23 more`

**configured the input directory:**
openapi.generator.input-base-dir=src/main/resources/openapi

ran ./gradlew build




### This happened:

`
Caused by: java.lang.StringIndexOutOfBoundsException: Range [0, -1) out of bounds for length 85
        at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:55)
        at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:52)
        at java.base/jdk.internal.util.Preconditions$4.apply(Preconditions.java:213)
        at java.base/jdk.internal.util.Preconditions$4.apply(Preconditions.java:210)
        at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:98)
        at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckFromToIndex(Preconditions.java:112)
        at java.base/jdk.internal.util.Preconditions.checkFromToIndex(Preconditions.java:349)
        at io.quarkiverse.openapi.server.generator.deployment.codegen.ApicurioOpenApiServerCodegen.lambda$getInputBaseDir$0(ApicurioOpenApiServerCodegen.java:43)
        at io.quarkiverse.openapi.server.generator.deployment.codegen.ApicurioOpenApiServerCodegen.getInputBaseDir(ApicurioOpenApiServerCodegen.java:41)
        at io.quarkiverse.openapi.server.generator.deployment.codegen.ApicurioOpenApiServerCodegen.trigger(ApicurioOpenApiServerCodegen.java:59)
        at io.quarkus.deployment.CodeGenerator.lambda$trigger$4(CodeGenerator.java:206)
        at io.quarkus.deployment.CodeGenerator.callWithClassloader(CodeGenerator.java:181)
        at io.quarkus.deployment.CodeGenerator.trigger(CodeGenerator.java:203)
        at io.quarkus.deployment.CodeGenerator.initAndRun(CodeGenerator.java:80)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        ... 23 more
`

### I expected this:

i expect that the server stubs are generated based on the yaml openapi spec

### Is there a workaround?

_No response_

### How can we try to reproduce the issue?

_No response_

### Anything else?

_No response_

### Output of `uname -a` or `ver`

Darwin mblxmjhjy2pv 23.5.0 Darwin Kernel Version 23.5.0: Wed May  1 20:12:58 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6000 arm64

### Output of `java -version`

OpenJDK Runtime Environment Temurin-21+35 (build 21+35-LTS)

### Quarkus OpenApi version or git rev

2.9.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 8.13

### Additional information

_No response_

### Community Notes

<!-- Please keep this note for the community -->
* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.name: Bug Report 🐞

> [@diegostalder](https://github.com/diegostalder) Do you would like to try to solve this one?

hmmm, i don't think i can do it. i guess i lack the basic knowledge needed.

My analysis/obeservation:

-> i tried it with a plain quarkus maven project. and it worked.

-> if i try it without configuring the base-dir. i get the following error:
Caused by: io.quarkus.bootstrap.prebuild.CodeGenException: The OpenAPI input base directory does not exist. Please create the directory at **/Users/sg/dev/diba/neo/eus-iam/build/classes/java/quarkus-generated-sources/resources**

what seems strange to me, that the directory points to the build/target direcory.
i would have expected something like "src/main/resources/openapi"

i guess it has something to to with the inputDir from CodeGenContext:

ApicurioOpenApiServerCodegen.java:
        final Path openApiDir = getInputBaseDir(**context.inputDir()**, context.config());

clearly, the getInputBaseDir() method expects a inputDir with "src" in it, hence the exception.

--- a/server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/ServerCodegenConfig.java
+++ b/server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/ServerCodegenConfig.java
@@ -7,6 +7,7 @@ import io.smallrye.config.WithDefault;
 public interface ServerCodegenConfig {
 
     String DEFAULT_PACKAGE = "io.apicurio.api";
+    String DEFAULT_DIR = "openapi";
 
     /**
      * The OpenAPI specification filename.

--- a/server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/codegen/ApicurioOpenApiServerCodegen.java
+++ b/server/deployment/src/main/java/io/quarkiverse/openapi/server/generator/deployment/codegen/ApicurioOpenApiServerCodegen.java
@@ -1,9 +1,12 @@
 package io.quarkiverse.openapi.server.generator.deployment.codegen;
 
+import static io.quarkiverse.openapi.server.generator.deployment.ServerCodegenConfig.DEFAULT_DIR;
+
 import java.io.File;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.Arrays;
+import java.util.Optional;
 
 import org.eclipse.microprofile.config.Config;
 import org.slf4j.Logger;
@@ -36,27 +39,34 @@ public class ApicurioOpenApiServerCodegen implements CodeGenProvider {
         return "resources";
     }
 
-    private Path getInputBaseDir(final Path sourceDir, final Config config) {
-        return config.getOptionalValue(CodegenConfig.getInputBaseDirPropertyName(), String.class)
-                .map(inputBaseDir -> {
-                    int srcIndex = sourceDir.toString().lastIndexOf("src");
-                    return Path.of(sourceDir.toString().substring(0, srcIndex), inputBaseDir);
-                }).orElse(Path.of(sourceDir.toString(), "openapi"));
+    private Optional<String> getInputBaseDirRelativeToModule(final Path sourceDir, final Config config) {
+        return config.getOptionalValue(CodegenConfig.getInputBaseDirPropertyName(), String.class).map(baseDir -> {
+            int srcIndex = sourceDir.toString().lastIndexOf("src");
+            return srcIndex < 0 ? null : Path.of(sourceDir.toString().substring(0, srcIndex), baseDir).toString();
+        });
     }
 
     @Override
     public boolean shouldRun(Path sourceDir, Config config) {
-        boolean specIsPresent = config.getOptionalValue(CodegenConfig.getSpecPropertyName(), String.class).isPresent();
-        if (!specIsPresent) {
+        Optional<String> possibleSpecPropertyName = config.getOptionalValue(CodegenConfig.getSpecPropertyName(), String.class);
+        if (possibleSpecPropertyName.isEmpty()) {
             log.warn("The {} property is not present, the code generation will be ignored",
                     CodegenConfig.getSpecPropertyName());
+            return false;
         }
-        return specIsPresent;
+        String specPropertyName = possibleSpecPropertyName.get();
+        String relativeInputBaseDir = getInputBaseDirRelativeToModule(sourceDir, config).orElse(null);
+        if (relativeInputBaseDir != null) {
+            return Files.exists(Path.of(relativeInputBaseDir).resolve(specPropertyName));
+        } else {
+            return Files.exists(sourceDir.resolve(DEFAULT_DIR).resolve(specPropertyName));
+        }
     }
 
     @Override
     public boolean trigger(CodeGenContext context) throws CodeGenException {
-        final Path openApiDir = getInputBaseDir(context.inputDir(), context.config());
+        final Path openApiDir = Path.of(getInputBaseDirRelativeToModule(context.inputDir(), context.config())
+                .orElse(context.inputDir().resolve(DEFAULT_DIR).toString()));
 
         validateOpenApiDir(context, openApiDir);
 

